modules/percolator/src/main/java/org/elasticsearch/percolator/PercolateQueryBuilder.java

```diff
@@ -557,41 +565,81 @@ static PercolateQuery.QueryStore createStore(MappedFieldType queryBuilderFieldTy
+                                    + "unavailable on index [{}].",
+                                context.index().getName()
+                            );
                         }
-                        // set the transportversion here - only read vints so far, so can change the version freely at this point
-                        input.setTransportVersion(transportVersion);
-
-                        QueryBuilder queryBuilder = input.readNamedWriteable(QueryBuilder.class);
-                        assert in.read() == -1;
-                        queryBuilder = Rewriteable.rewrite(queryBuilder, context);
-                        return queryBuilder.toQuery(context);
-                    }
-
+                        LOGGER.warn(
+                            "Reading percolator query from source. For best performance, reindexing of index [{}] is required.",
+                            context.index().getName()
+                        );
+                        SourceProvider sourceProvider = context.createSourceProvider(new SourceFilter(null, null));
+                        Source source = sourceProvider.getSource(ctx, docId);
+                        SourceToParse sourceToParse = new SourceToParse(
+                            String.valueOf(docId),
+                            source.internalSourceRef(),
+                            source.sourceContentType()
+                        );
+
+                        return context.parseDocument(sourceToParse).rootDoc().getBinaryValue(queryBuilderFieldType.name());
+                    });
+
+                    queryBuilder = Rewriteable.rewrite(queryBuilder, context);
+                    return queryBuilder.toQuery(context);
                 } else {
                     return null;
                 }
             };
         };
     }
 
+    private static QueryBuilder readQueryBuilder(
+        BytesRef bytesRef,
+        NamedWriteableRegistry registry,
+        IndexVersion indexVersion,
+        CheckedSupplier<BytesRef, IOException> fallbackSource
+    ) throws IOException {
+        try (
+            InputStream in = new ByteArrayInputStream(bytesRef.bytes, bytesRef.offset, bytesRef.length);
+            StreamInput input = new NamedWriteableAwareStreamInput(new InputStreamStreamInput(in, bytesRef.length), registry)
+        ) {
+            // Query builder's content is stored via BinaryFieldMapper, which has a custom encoding
+            // to encode multiple binary values into a single binary doc values field.
+            // This is the reason we need to first read the number of values and
+            // then the length of the field value in bytes.
+            int numValues = input.readVInt();
+            assert numValues == 1;
+            int valueLength = input.readVInt();
+            assert valueLength > 0;
+
+            TransportVersion transportVersion;
+            if (indexVersion.before(IndexVersions.V_8_8_0)) {
```

Comment:
I assume when we remove all versions prior to V_8_8_0 this if/else just becomes the else in a safe way now since it'll either reload from source or reject if there is no source?